src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,312 @@
+                    // it exited early so we looped around and DateTime.Now=20.99.
+                    // Aligning up from DateTime.Now would give us 21.00 again so we also need to skip
+                    // forward one time interval
+                    DateTime minNextInterval = intervalStartTime.AddSeconds(collectionIntervalSecs);
+                    if (nextIntervalStartTime <= minNextInterval)
+                    {
+                        nextIntervalStartTime = minNextInterval;
+                    }
+
+                    // pause until the interval is complete
+                    TimeSpan delayTime = nextIntervalStartTime - now;
+                    cancelToken.WaitHandle.WaitOne(delayTime);
+
+                    // don't do collection if timer may not have run to completion
+                    if (cancelToken.IsCancellationRequested)
+                    {
+                        break;
+                    }
+
+                    // collect statistics for the completed interval
+                    _beginCollection(intervalStartTime);
+                    Collect();
+                    _endCollection(intervalStartTime);
+                    intervalStartTime = nextIntervalStartTime;
+                }
+            }
+            catch (Exception e)
+            {
+                _collectionError(e);
+            }
+        }
+
+        public void Dispose()
+        {
+            _cts.Cancel();
+            if (_collectThread != null)
+            {
+                _collectThread.Join();
+                _collectThread = null;
+            }
+            _listener.Dispose();
+        }
+
+        private void RemoveInstrumentState(Instrument instrument, InstrumentState state)
+        {
+            _instrumentStates.TryRemove(instrument, out InstrumentState? _state);
+        }
+
+        private InstrumentState? GetInstrumentState(Instrument instrument)
+        {
+            if (!_instrumentStates.TryGetValue(instrument, out InstrumentState? instrumentState))
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
+                {
+                    foreach (var filter in _instrumentConfigFuncs)
+                    {
+                        if (filter(instrument))
+                        {
+                            instrumentState = BuildInstrumentState(instrument);
+                            break;
+                        }
+                    }
+                }
+                if (instrumentState != null)
+                {
+                    _instrumentStates.TryAdd(instrument, instrumentState);
+                    instrumentState = _instrumentStates[instrument];
+                }
+            }
+            return instrumentState;
+        }
+
+        internal InstrumentState? BuildInstrumentState(Instrument instrument)
+        {
+            Func<Aggregator>? createAggregatorFunc = GetAggregatorFactory(instrument);
+            if (createAggregatorFunc == null)
```

Comment:
>createAggregatorFunc == null [](http://example.com/codeflow?start=16&length=28)

nit: do you prefer this syntax or you prefer using 
if (x is null) and if (x is not null)?

Reply:
My personal preference is x == null but I'm happy to do it differently if there is a general convention or consensus that 'is null' is the better one?

Reply:
It is not a big deal. I used the other style in the Metrics code. this is why I asked about it :-)